pipit/trace.py

```diff
@@ -681,3 +681,65 @@ def calc_exc_time_in_bin(events):
         df.insert(0, "bin_end", edges[1:])
 
         return df
+
+    def detect_pattern(
+        self,
+        start_event,
+        iterations=None,
+        window_size=None,
+        process=0,
+        metric="time.exc",
+    ):
+        import stumpy
+
+        # count the number of enter events to
+        # determine the number of iterations if it's not
+        # given by the user.
+        if iterations is None:
+            iterations = len(
+                self.events[
+                    (self.events["Name"] == start_event)
+                    & (self.events["Event Type"] == "Enter")
+                    & (self.events["Process"] == process)
+                ]
+            )
+
+        # get the first enter and last leave of
+        # the given event. we will only investigate
+        # this portion of the data.
+        first_loop_enter = self.events[
+            (self.events["Name"] == start_event)
+            & (self.events["Event Type"] == "Enter")
+            & (self.events["Process"] == process)
+        ].index[0]
+
+        last_loop_leave = self.events[
+            (self.events["Name"] == start_event)
+            & (self.events["Event Type"] == "Leave")
+            & (self.events["Process"] == process)
+        ].index[-1]
+
+        df = self.events.iloc[first_loop_enter + 1 : last_loop_leave]
```

Comment:
any reason why we don't include first_loop_enter and last_loop_leave themselves?

Reply:
I don't remember but I will think about it and add comments